5 4 customicon references

smeyer-ilias opened the pull request:
@mjansenDatabay This is an implementation for the custom icon presenter for the general presention (limited to the methods exists() and getAbsolutePath()) for container references.

smeyer-ilias committed:
Added presenter for custom icons

smeyer-ilias committed:
Usage for custom icon presenter

mjansenDatabay commented on Services/Object/Icon/classes/class.ilObjectCustomIconFactory.php:
```diff
@@ -78,4 +91,36 @@ public function getByObjId($objId, $objType = '')
+			case 'catr':
+			case 'grpr':
+			case 'crsr':
+				$target_obj_id = \ilObjectReferenceCustomIconPresenter::lookupTargetId($objId, $this->dbInterface);
```
Do you think we can get rid of the static method call here?

We already have a lookup method here: `ilContainerReference::_lookupTargetId($a_obj_id);`

I suggest to inject some class as a dependency to the constructor of this factory, which internally wraps this lookup and can be used here.

smeyer-ilias replied:
@mjansenDatabay You are absolutely right, the method _lookupTargetId should be reused. 
Regarding your suggestion to avoid the static method, I see three possibilities: 

- as you said add the "ContainerReference" as dependency and determine the target ref_id. But after this is done, an instance of ilObjectCustomIconImpl needs to be created and this would require to add the factory itself as dependency for the "ilObjectReferenceCustomIconPresenter". => something I like to avoid

- keep the static method to look up the target id and then create a ilObjectCustomIconPresenter without further dependencies.

- get rid of the static method and directly call ilContainerReference::_lookupTargetId in the factory for container references.

mjansenDatabay replied:
@smeyer-ilias As you already wrote the dependency could be passed to the factory constructor AND delegated to the constructor of `ilObjectReferenceCustomIconPresenter` if an instance is created.

If you want to avoid this, I prefer option 3.

mjansenDatabay commented on Services/Object/Icon/classes/class.ilObjectReferenceCustomIconPresenter.php:
```diff
@@ -0,0 +1,62 @@
+	 * @return int
+	 * @throws ilDatabaseException
+	 */
+	public static function lookupTargetId($a_obj_id, \ilDBInterface $db)
```
As written in my comment above, can be removed.

mjansenDatabay commented:
@smeyer-ilias IMO there are only tiny changes required, maybe you could additionally introduce PHP 7 type hints and return type declarations, but this is only sugar on top ;-).

mjansenDatabay commented on Services/Object/Icon/classes/class.ilObjectCustomIconFactory.php:
```diff
@@ -20,17 +20,30 @@ class ilObjectCustomIconFactory
+		\ILIAS\Filesystem\Filesystem $webDirectory,
+		\ILIAS\FileUpload\FileUpload $uploadService,
+		\ilObjectDataCache $objectCache,
+		\ilDBInterface $db
```
Not required if a "lookup service" for resolving the target object is injected instead.

smeyer-ilias committed:
Added new dependency for ReferencePresenter; removed DB dependency for IconFactory; Added parameter/return type hints

smeyer-ilias commented:
Added new version with factory dependency. 
Furthermore I added parameter and return type hints and noticed that some of you methods hinted the implementation classes and not the interfaces. Let me know if this was intentionally.

mjansenDatabay commented:
@smeyer-ilias LGTM! Thanks a lot! Hinting concrete implementations was done by mistake.

mjansenDatabay commented:
Should I merge this, or does @alex40724 have any objections?

alex40724 commented:
No objections from my side.
